### PR TITLE
feature #842: Implemented an endpoint to receive all disciplines

### LINKED ISSUE #66: Implement an endpoint to receive all disciplines
Your task is to create a common endpoint for _receiving all disciplines_. Follow the design and guidelines below.

image

- **Pagination**
- **Sorting**
  - by name
  - by year and semester
  - by group
- **Filtering**
  - by groups _(several groups)_ 
  - by years and semesters 
  ```ts
  [{ year: 1939, semester: 1 }, { year: 1939, semester: 2 } ...]
  ```
  - by teachers _(not discipline teachers)_

### LINKED ISSUE #842: Implement an endpoint to receive all disciplines
Your task is to create a common endpoint for _receiving all disciplines_. Follow the design and guidelines below.

image

- **Pagination**
- **Sorting**
  - by name
  - by year and semester
  - by group
- **Filtering**
  - by groups _(several groups)_ 
  - by years and semesters 
  ```ts
  [{ year: 1939, semester: 1 }, { year: 1939, semester: 2 } ...]
  ```
  - by teachers _(not discipline teachers)_

### PR DESCRIPTION
**Pull request by JessFreak**
- created getStrictSearch in DatabaseUtils
- created getAll in DisciplineController
- created getAll in DisciplineService
- created getDisciplinesForAdmin in DisciplineMapper
- created count in DisciplineRepository
- created DisciplineAdminResponse and DisciplinesResponse
- created QueryAllDisciplinesDTO
- created QueryAllDisciplinesPipe
- created ShortTeacherResponse

closes fictadvisor/fictadvisor#66

### CODE REVIEW

File: src/v2/api/responses/DisciplineResponse.ts [hunk 1]
```diff
@@ -22,6 +25,34 @@ export class BaseDisciplineResponse {
   })
     semester: number;
 }
+
+export class DisciplineAdminResponse extends BaseDisciplineResponse {
+  @ApiProperty({
+    description: 'Group\' associated with the discipline',
```
Comment (f-ogoog) [inline src/v2/api/responses/DisciplineResponse.ts:31]: needless \'

File: src/v2/api/responses/DisciplineResponse.ts [hunk 2]
```diff
@@ -22,6 +25,34 @@ export class BaseDisciplineResponse {
...
+  })
+    group: GroupResponse;
+
+  @ApiProperty({
+    description: 'Teacher\' associated with the discipline',
```
Comment (f-ogoog) [inline src/v2/api/responses/DisciplineResponse.ts:36]: needless '

File: src/v2/api/services/DisciplineService.ts [hunk 3]
```diff
@@ -14,6 +19,36 @@ export class DisciplineService {
     private disciplineTeacherRepository: DisciplineTeacherRepository,
   ) {}
 
+  private DisciplineSearching = {
+    name: (name: string) => ({
```
Comment (f-ogoog) [inline src/v2/api/services/DisciplineService.ts:23]: rename parameter name to search

File: src/v2/api/services/DisciplineService.ts [hunk 4]
```diff
@@ -22,6 +57,30 @@ export class DisciplineService {
...
+    const data = {
+      where: {
+        AND: [
+          this.DisciplineSearching.name(name),
+          this.DisciplineSearching.groups(groups),
+          this.DisciplineSearching.semesters(semesters),
+          this.DisciplineSearching.teachers(teachers),
+        ],
+      },
+      orderBy: this.DisciplineSorting[sort](order),
```
Comment (f-ogoog) [inline src/v2/api/services/DisciplineService.ts:79]: use this function in variable before put it in data

File: src/v2/database/repositories/DisciplineRepository.ts [hunk 5]
```diff
@@ -73,4 +73,10 @@ export class DisciplineRepository {
   async updateMany (where: Prisma.DisciplineWhereInput, data: Prisma.DisciplineUncheckedUpdateManyInput) {
     return this.prisma.discipline.updateMany({ where, data });
   }
+
+  async count (data: Prisma.DisciplineCountArgs): Promise<number> {
+    return this.prisma.discipline.count(
+      data,
+    );
```
Comment (f-ogoog) [inline src/v2/database/repositories/DisciplineRepository.ts:80]: needless enters

**Review by f-ogoog — CHANGES_REQUESTED** [review]

File: src/v2/api/responses/DisciplineResponse.ts [hunk 6]
```diff
@@ -22,6 +25,34 @@ export class BaseDisciplineResponse {
   })
     semester: number;
 }
+
+export class DisciplineAdminResponse extends BaseDisciplineResponse {
+  @ApiProperty({
+    description: 'Group associated with the discipline',
```
Comment (f-ogoog) [inline src/v2/api/responses/DisciplineResponse.ts:31]: add Type

File: src/v2/api/dtos/QueryAllDisciplinesDTO.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
+import { ApiPropertyOptional } from '@nestjs/swagger';
+import { IsIn, IsOptional, ValidateNested } from 'class-validator';
+import { QueryAllDTO } from '../../utils/QueryAllDTO';
+import { QuerySemesterDTO } from './QuerySemesterDTO';
+import { validationOptionsMsg } from '../../utils/GLOBALS';
+
+export class QueryAllDisciplinesDTO extends QueryAllDTO {
+  @ApiPropertyOptional({
+    description: 'Search disciplines by name',
+  })
```
Comment (f-ogoog) [inline src/v2/api/dtos/QueryAllDisciplinesDTO.ts:10]: add IsOptional

**Review by f-ogoog — CHANGES_REQUESTED** [review]

**Review by f-ogoog — APPROVED** [review]

File: src/v2/api/responses/DisciplineResponse.ts [hunk 8]
```diff
@@ -22,6 +25,35 @@ export class BaseDisciplineResponse {
...
+    type: GroupResponse,
+  })
+    group: GroupResponse;
+
+  @ApiProperty({
+    description: 'Teachers associated with the discipline',
+    type: [ShortTeacherResponse],
+  })
+    teachers: ShortTeacherResponse[];
+}
```
Comment (SharpDevOps10) [inline src/v2/api/responses/DisciplineResponse.ts]: **Disciplines**, an array of objects, has `name` field. You missed it => add it to your response

File: src/v2/api/dtos/QueryAllDisciplinesDTO.ts [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+  @IsOptional()
+    search?: string;
```
Comment (SharpDevOps10) [inline src/v2/api/dtos/QueryAllDisciplinesDTO.ts]: You've extended `QueryAllDTO` with `search` field => there is no need to add this filed to `QueryAllDisciplinesDTO`

**Review by SharpDevOps10 — CHANGES_REQUESTED** [review]
Also filtering using `semesters` does not work :(

**Review by SharpDevOps10 — APPROVED** [review]

File: src/v2/api/dtos/QueryAllDisciplinesDTO.ts [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+  @IsOptional()
+    groups?: string[];
+
+  @ApiPropertyOptional({
+    description: 'Array of studying semesters',
+  })
+  @IsOptional()
+  @ValidateNested({
+    each: true,
+  })
```
Comment (fokaaas) [inline src/v2/api/dtos/QueryAllDisciplinesDTO.ts:21]: Add ```@Type(() => QuerySemesterDTO)```

**Review by fokaaas — CHANGES_REQUESTED** [review]
So, test it through axios and it will work :)

**Comment by fokaaas** [comment]
closed as outdated

**Pull request closed without merging**